api/src/oscc.c

```diff
@@ -492,110 +615,366 @@ oscc_result_t oscc_async_enable( int socket )
+
+              vehicle_detection.has_wheel_speed |=
+                  ( rx_frame.can_id == KIA_SOUL_OBD_WHEEL_SPEED_CAN_ID );
         }
     }
 
+    close( sock );
 
-    if ( result == OSCC_OK )
+    can_contains_s detection =
     {
-        /* all prior checks will pass even if a valid interface has not been
-           set up - attempt to write an empty CAN frame to the interface to see
-           if it is valid */
-        struct can_frame tx_frame;
+      .is_oscc = oscc_detection.has_brake_report &&
+                 oscc_detection.has_steer_report &&
+                 oscc_detection.has_accel_report,
+      .has_vehicle = vehicle_detection.has_brake_pressure &&
+                     vehicle_detection.has_steering_angle &&
+                     vehicle_detection.has_wheel_speed
+    };
 
-        memset( &tx_frame, 0, sizeof(tx_frame) );
-        tx_frame.can_id = 0;
-        tx_frame.can_dlc = 8;
+    return detection;
+}
 
-        int bytes_written = write( sock, &tx_frame, sizeof(tx_frame) );
 
-        if ( bytes_written < 0 )
-        {
-            printf( "Failed to write test frame to %s: %s\n", can_channel, strerror(errno) );
+oscc_result_t construct_interfaces_list( device_names_s * const names_ptr )
+{
+    FILE *file_handler;
+    char buffer[512];
+    oscc_result_t result = OSCC_OK;
 
-            result = OSCC_ERROR;
-        }
-        else
-        {
-            can_socket = sock;
+    file_handler = fopen( "/proc/net/dev", "r" );
+    if (!file_handler) {
+        perror( "Cannot read: /proc/net/dev" );
+
+        return OSCC_ERROR;
+    }
+
+    int lines = 0;
+
+    while ( fgets( buffer, sizeof( buffer ), file_handler ) ) {
+        ++lines;
+    }
+
+    names_ptr->name = (char**) malloc(lines * sizeof(char*));
+
+    int i;
+
+    for ( i = 0; i < lines; i++ )
+    {
+        names_ptr->name[i] = (char*) malloc(IFNAMSIZ * sizeof(char));
+    }
+
+    rewind( file_handler );
+
+    //Consume the first two lines since they are headers
+    fgets( buffer, sizeof( buffer ), file_handler );
+    fgets( buffer, sizeof( buffer ), file_handler );
+
+    int size = 0;
+
+    char* socket_name = calloc( IFNAMSIZ, sizeof(char) );
+
+    if( !socket_name )
+    {
+        result = OSCC_ERROR;
+    }
+    else
+    {
+        while ( size < lines && fgets( buffer, sizeof( buffer ), file_handler ) ) {
+
+            get_device_name( buffer, socket_name );
+
+            strncpy( names_ptr->name[size], socket_name, IFNAMSIZ );
+
+            size++;
         }
+
+        free( socket_name );
+
+        names_ptr->size = size;
     }
 
+    fclose( file_handler );
 
     return result;
 }
+
+void clear_device_names( device_names_s * const names_ptr )
+{
+    int i;
+
+    for ( i = 0; i < names_ptr->size; i++ )
+    {
+        free( names_ptr->name[i] );
+    }
+
+    free( names_ptr->name );
+}
+
+
+void get_device_name( char * string, char * const name )
```

Comment:
if `name` isn't exactly what you're expecting this will crash. Minimally a check for `NULL` before writing to it?

Reply:
Looks like there are more instances above where pointer params that aren't checked for NULL. Is that an intentional omission?